### PR TITLE
Enable support for externally-defined ID generators

### PR DESCRIPTION
**Pull request by Aneurysm9**
* Moved the SDK's `internal.IDGenerator` interface to the `sdk/trace`
package.
* Added `trace.WithIDGenerator()` `TracerProviderOption`.

### CODE REVIEW

**Review by XSAM — APPROVED** [review]

File: sdk/trace/id_generator.go [hunk 1]
```diff
@@ -15,20 +15,26 @@
...
+	"encoding/binary"
 	"math/rand"
 	"sync"
 
 	"go.opentelemetry.io/otel/trace"
-
-	"go.opentelemetry.io/otel/sdk/trace/internal"
 )
 
+// IDGenerator allows custom generators for TraceId and SpanId.
```
Comment (krnowak) [inline sdk/trace/id_generator.go:26]:
```suggestion
// IDGenerator allows custom generators for TraceID and SpanID.
```

File: sdk/trace/id_generator.go [hunk 2]
```diff
@@ -15,20 +15,26 @@
...
+	NewSpanID() trace.SpanID
+}
+
 type defaultIDGenerator struct {
 	sync.Mutex
 	randSource *rand.Rand
 }
 
-var _ internal.IDGenerator = &defaultIDGenerator{}
+var _ IDGenerator = &defaultIDGenerator{}
```
Comment (krnowak) [inline sdk/trace/id_generator.go:37]: How about renaming the `defaultIDGenerator` type to `randomIDGenerator`, and then renaming the `defIDGenerator` function as `defaultIDGenerator`?

**Review by krnowak — CHANGES_REQUESTED** [review]

**Review by krnowak — APPROVED** [review]

**Comment by jmacd** [comment]
I made a note in the associated spec PR (https://github.com/open-telemetry/opentelemetry-specification/pull/1006#discussion_r520693556) that I'd probably rather see two methods, where one returns a single SpanID (taking a mutex once) and one returns a pair of TraceID and SpanID (taking a mutex once). What do you think?

**Comment by krnowak** [comment]
> I made a note in the associated spec PR ([open-telemetry/opentelemetry-specification#1006 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/1006#discussion_r520693556)) that I'd probably rather see two methods, where one returns a single SpanID (taking a mutex once) and one returns a pair of TraceID and SpanID (taking a mutex once). What do you think?

Sounds useful as this seems to be how the ID generator is going to be used. Would there be a case where we only wanted to generate a trace ID?

**Comment by MrAlias** [comment]
> I made a note in the associated spec PR ([open-telemetry/opentelemetry-specification#1006 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/1006#discussion_r520693556)) that I'd probably rather see two methods, where one returns a single SpanID (taking a mutex once) and one returns a pair of TraceID and SpanID (taking a mutex once). What do you think?

Yeah, I think this is indeed going to be how this will be used.

> Would there be a case where we only wanted to generate a trace ID?

I feel like that would be the corner case anyway so throwing away that returned value wouldn't be too much of an issue.

File: sdk/trace/id_generator.go [hunk 3]
```diff
@@ -15,23 +15,29 @@
...
 	"go.opentelemetry.io/otel/trace"
-
-	"go.opentelemetry.io/otel/sdk/trace/internal"
 )
 
-type defaultIDGenerator struct {
+// IDGenerator allows custom generators for TraceID and SpanID.
+type IDGenerator interface {
+	NewTraceID() trace.TraceID
+	NewSpanID() trace.SpanID
```
Comment (MrAlias) [inline sdk/trace/id_generator.go:29]: I like @jmacd's idea of having these be a function that returns a span ID and a trace ID with a span ID. I also wonder if we can include arguments that would be helpful for ID generation. Maybe something like:

```suggestion
	NewIDs(ctx context.Context) (trace.TraceID, trace.SpanID)
	NewSpanID(ctx context.Context, traceID trace.TraceID) trace.SpanID
```

I think passing a context will enable generation with the context of things like baggage or other vendor defined values. Also, passing the trace ID to the NewSpanID method might be a useful things if a generator uses that in relation to the span IDs it generates.

**Review by dashpole — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

File: sdk/trace/id_generator.go [hunk 4]
```diff
@@ -15,36 +15,53 @@
...
 	sid := trace.SpanID{}
 	gen.randSource.Read(sid[:])
 	return sid
 }
 
-// NewTraceID returns a non-zero trace ID from a randomly-chosen sequence.
-// mu should be held while this function is called.
-func (gen *defaultIDGenerator) NewTraceID() trace.TraceID {
+// NewIDs returns a non-zero trace ID and a non-zero span ID from a
+// randomly-chosen sequence. mu should be held while this function is called.
```
Comment (krnowak) [inline sdk/trace/id_generator.go:50]: I'd drop the comment about `mu`. Mutex is now embedded, but I don't think that anyone should care about this implementation detail.

```suggestion
// randomly-chosen sequence.
```

**Review by krnowak — APPROVED** [review]
One nit, otherwise looks good.

**Pull request merged**